### PR TITLE
Fixing sparse linear race condition

### LINKED ISSUE #790: SparseLinear:accGradParameters wrong multithreaded behavior?
#789 is a test that passes with OMP_NUM_THREADS=1, but not otherwise with OMP + OpenBlas (or MKL, for that matter). The OMP statement in the SparseLinear.c accGradParameters function makes concurrent updates of gradWeight possible if the same index occurs multiple times in the input. The legacy code should be wrong too. Anything I missed?

EDIT: the behavior we get out of this will make the algo converge eventually as it amounts to doing "hogwild" updates. But it should at least be an option.

### PR DESCRIPTION
**Pull request by ebetica**
Fixes #790 and #722

### CODE REVIEW

**Comment by soumith** [comment]
Thanks Zeming!

File: lib/THNN/generic/SparseLinear.c [hunk 1]
```diff
@@ -59,25 +59,43 @@ void THNN_(SparseLinear_updateOutput)(
 
   long nnz = THTensor_(size)(input, 0);
 
+  THLongTensor * csr = THLongTensor_newWithSize1d(batchSize+1);
```
Comment (fmassa) [inline lib/THNN/generic/SparseLinear.c]: Isn't this tensor leaking ?

File: SparseLinear.lua [hunk 2]
```diff
@@ -130,8 +130,13 @@ function SparseLinear:accGradParameters(input, gradOutput, scale)
          gradOutput:resize(1, gradOutput:size(1))
       end
 
+      local rows = self.formatted_input:select(2, 1)
+      local cols = self.formatted_input:select(2, 2)
+      local sortinds = cols * gradOutput:size(1) + rows
```
Comment (fmassa) [inline SparseLinear.lua]: This line and the following 2 allocates memory at each call

File: lib/THNN/generic/SparseLinear.c [hunk 3]
```diff
@@ -165,26 +183,42 @@ void THNN_(SparseLinear_accGradParameters)(
              "gradOutput must be contiguous");
 
   long nnz = THTensor_(size)(input, 0);
-  // THTensor_(resize2d)(gradOutput, batchSize, outDim);
 
-  // gradWeight += gradOutput * input
-#pragma omp parallel for private(h, i) schedule(static) if (\
-  nnz * outDim > 10000)
+  THLongTensor* csc = THLongTensor_newWithSize1d(inDim+1);
```
Comment (fmassa) [inline lib/THNN/generic/SparseLinear.c]: This tensor needs to be freed as well, right?

**Comment by fmassa** [comment]
Hi,
Sorry for my lack of reactivity those past few weeks, I've been quite busy lately.
I made some inline comments.
Thanks for the fix !

File: lib/THNN/generic/SparseLinear.c (hunk 3 above)
Comment (soumith) [inline lib/THNN/generic/SparseLinear.c]: yea both are n00b leaks. I'm fixing them. cc: @ebetica

**Pull request merged**